contracts/rules/CreditRule.sol

```diff
@@ -0,0 +1,192 @@
+pragma solidity ^0.5.0;
+
+// Copyright 2018 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "../external/SafeMath.sol";
+import "../token/TransfersAgent.sol";
+
+
+/**
+ * Credit rules allows a budget holder to credit a user with an amount
+ * that the user can *only* spents in a token economy. A custom rule
+ * is deployed with a CreditRule acting as a TransfersAgent to execute
+ * transfers. CreditRule first transfers a minimum of credited amount and
+ * needed transfer amount, afterwards delegate the transfer execution to
+ * TransfersAgent itself.
+ *
+ * Steps to utilize CreditRule:
+ *  - A custom rule is deployed by passing a CreditRule address to act as
+ *    a transfers agent for the rule.
+ *  - A token holder signs an executable transaction to execute the custom rule.
+ *  - The budget holder signs an executable transaction to execute
+ *    CreditRule::executeRule function with a credit amount, and the token
+ *    holder's address and executeRule's function data as an argument.
+ */
+contract CreditRule is TransfersAgent {
+
+    /* Usings */
+
+    using SafeMath for uint256;
+
+
+    /** Structs */
+
+    struct CreditInfo {
+        uint256 amount;
+        bool inProgress;
+    }
+
+
+    /* Storage */
+
+    address public budgetHolder;
+
+    TransfersAgent public transfersAgent;
+
+    mapping(address => CreditInfo) public credits;
```

Comment:
Do we need to handle use case where credits for a user are only valid for a particular duration or blockheight? 💭 
I might not be aware of scope of the project.